h2o-core/src/main/java/hex/grid/filter/DefaultPermutationFilter.java

```diff
@@ -0,0 +1,38 @@
+package hex.grid.filter;
+
+import hex.Model;
+
+import java.util.List;
+
+public class DefaultPermutationFilter<MP extends Model.Parameters> implements PermutationFilter<MP> {
```

Comment:
why is this not simply itself a `PermutationFilterFunction`?
it's creates too many level of filters and it's confusing.
I think everything should be named `Filter` for the sake of simplicity.
No need to add another level with a negative semantic:`filter.permutationIsSkipped(..)` vs `filter.test(..)`.
That's the main benefit of defining your filter interface.
And it's only when you know that you want to use this filter to skip some combinations that you apply the negation:
```
if(! _permutationFilter.test(nextMPs)) {
    // try next
} else {
    // use this one
}
```

Reply:
Also would then recommend to rename it sth like `AllMatchFilter` to make clear that all the filters passed to the constructor of this one should match to make this one match.

Reply:
@seb-h2o 
> why is this not simply itself a PermutationFilterFunction?
it creates too many level of filters and it's confusing.

Will try to fix this. IMO These things are slightly different. PermutationFilterFunctions decides whether permutation should be considered for being filtered out. DefaultPermutationFilter decides what we should do with permutation given results from all underneath filter functions - so it is not only testing.

Reply:
@seb-h2o I think that word `test` does not have enough connotation to make it clear whether `_permutationFilter.test` should return true or false to filter permutation out. So I'm not convinced that it will make things less confusing.

Reply:
I suggest maybe `permutationIsSkipped` -> `skip`? wdyt?

Reply:
I don't like `skip` either tbh.
My main point is that in most APIs filter are just selectors, it's a function applying a predicate to a sequence.:
> Java Stream filter: Returns a stream consisting of the elements of this stream that match the given predicate.
> Python builtin filter: Construct an iterator from those elements of iterable for which function returns true.
> ...

So giving this `DefaultPermutationFilter` a negative semantic like `skip` makes it more difficult to reason with.
In my opinion, it's what you do with the result that should tell if you are filtering in or filtering out (skipping), not the filter itself. That's why I suggest you to move the negation where you actually apply the filter.

> I think that word test does not have enough connotation

That's what I like about it. 
>  PermutationFilterFunctions decides whether permutation should be considered for being filtered out. DefaultPermutationFilter decides what we should do with permutation given results from all underneath filter functions - so it is not only testing.

You seem to apply at top-down approach here, so that every component semantic is driven by your final goal. For this kind of logic, I think the bottom-up approach is much better: you compose simple components (basic filters) that you can reuse, combine, compose larger filters with smaller ones... and only when you're using them, they acquire a more specific semantic.

Reply:
@seb-h2o > My main point is that in most APIs filter are just selectors, it's a function applying a predicate to a sequence.

in most APIs filters are stateless. We would not have this conversation without need for stateful one.

> You seem to apply at top-down approach here, so that every component semantic is driven by your final goal. For this kind of logic, I think the bottom-up approach is much better: you compose simple components (basic filters) that you can reuse, combine, compose larger filters with smaller ones... and only when you're using them, they acquire a more specific semantic.

this is my goal as well. This is not I believe top-down or bottom-up. This is about separation of concerns. When you specify function you just care about its own semantic. This is imo an easier  way for users to specify what they want to filter out. 

`and only when you're using them, they acquire a more specific semantic` - don't see that it contradicts with what I'm doing. We are using functions from `DefaultPermutationFilter` and that is where they acquire extra semantic.

Reply:
`permutationIsSkipped` is extra semantic that has nothing to do with filtering